File: 3rd_Year/ca341/Assigment/euler/problems/euler-119-digit-power-sum.go/euler-119-digit-power-sum.go
package main

/*
Import needed packages for solving the problem at hand
fmt -> used for printing the output
math -> used for handling the math operations
strconv -> used for converting strings to integers
os -> used for getting command line input
*/

import (
	"fmt"
	"math"
	"strconv"
	"os"
)

/*
Function below takes one argument, an integer that
is converted to a string for each digit of the number
to check that the sum of the digits raised to some power
returns the original number rerutning the number if so
else returns -1
*/
func check(n int) int {
	num := strconv.Itoa(n)
	tot := 0

	//for loop that adds each digit handling ascii value conversion
	for _, i := range num {
		tot += int(i - '0')
	}
	if n%tot == 0 {
		j := 0
		//for loop to raise the digit to some power from 0 to 10
		for math.Pow(float64(tot), float64(j)) != float64(n) && j < 10 {
			j++
		}
		//return the number if its equal to the number passed
		if math.Pow(float64(tot), float64(j)) == float64(n) {
			return n
		}
	}

	return -1
}

func main() {
	var nums []int
	i := 1

	//find the nth number imnputed
	limit := os.Args[1]
    limit1, _ := strconv.Atoi(limit)

	//-> for loop that gets a list of numbers that satisfy the conditions
	//-> of the function above and gets a list of these numbers of a length specified
	//-> by the user
	for len(nums) != limit1 {
		result := check(i)
		if result != -1 {
			//start at 80 as a1 = 81
			if i > 80 {
				nums = append(nums, i)
			}
		}
		i++
	}

	//print the last elemnt of the list or otherwise know as the
	//number specified by the user on the command line
	fmt.Printf("%d\n", nums[limit1-1])
	
}

File: 3rd_Year/ca341/Assigment/euler/problems/euler-119-digit-power-sum.kt/euler-119-digit-power-sum.kt
import kotlin.math.pow

class Problem119 {
    //class mathod to check for the sum of the digits
    //being equal to a number raised for some power
    //Even a30 does not surpass a power of 10
    fun check(n: Int): Int {
        val num = n.toString()
        var tot = 0

        for (i in num) {
            tot += i.toString().toInt()
        }

        if (n % tot == 0) {
            var j = 0
            while (tot.toDouble().pow(j) != n.toDouble() && j < 10) {
                j++
            }
            if (tot.toDouble().pow(j) == n.toDouble()) {
                return n
            }
        }
        return -1
    }

    //uses the above function to find elements within the range provided
    fun cal(n: Int): List<Int> {
        val nums = mutableListOf<Int>()
        var i = 1

        while (nums.size != n) {
            val result = check(i)
            if (result != -1) {
                if (i > 80) {
                    nums.add(i)
                }
            }
            i++
        }

        return nums
    }
}

fun main(args: Array<String>) {
    val limit = Integer.parseInt(args[0])
    val problem = Problem119()
    val nums = problem.cal(limit)


    println(nums[limit-1])
}
